### ISSUE TITLE
Native endianess in TPR files

### ISSUE THREAD

**Issue by ezavod** [MDAnalysis/mdanalysis#2428]
This is probably an upstream problem, so this mostly aims at MDAnalysis' docs:

[Here](https://github.com/MDAnalysis/mdanalysis/blob/4adbd6c01673fcc9d13d4cd5c9089c1139257a57/package/MDAnalysis/topology/tpr/utils.py#L133-L136) it says that the new body is little-endian. I guess that's not correct.
In [tpr_test.zip](https://github.com/MDAnalysis/mdanalysis/files/4000298/tpr_test.zip) one could find `n3_le.tpr` and `n3_be.tpr` which are created on a little- and big-endian system respectively. Those could be inspected with `xxd n3_be.tpr`.
The header (until offset `0000:0063`), done with the original XDR serializer, is the same. But hereafter the new in memory serializer takes over and the files differ. The order of the bytes (i.e. padded 4 byte blocks) is reverse. GROMACS itself can't read big-endian files on a little-endian system and vice versa.
So it's native byte order instead of the documented little-endian.

There seem to be [efforts](http://redmine.gromacs.org/projects/gromacs/repository/revisions/ed91f1f7032b1411d33c254165d4cf6836ef318f/diff/src/gromacs/utility/inmemoryserializer.cpp) to enable endian swapping, but I don't know if this will apply to the TPR body and if the `standard` will be set to big-endian or little-endian.
Should one raise an issue in GROMACS regarding the TPR body?

PS: The comment goes on and explains the padding to four bytes. Curiously at some places they are not zero padded but one padded. E.g. Following the aforementioned header there are the box dimension. Those are in version 2019.4 `40 80 00 00` and in the new version (2020, rc1) `00 00 00 40  FF FF FF 80  0...`. I have yet to figure out where those Fs are coming from (and why this padding is even nescesary).

ping @jbarnoud

**Comment by jbarnoud** [comment]
Oh! Thank you for this comment! Indeed, based on how the TPR body ends up little endian (packing each byte as a "char") it makes sense that it is native endian, I just did not realize that as I only worked on this on little-endian systems.

Is there a way of knowing what is the endianness of the TPR body? Ideally we want to be able to read TPR generated on any endianness, regardless of the native endianness. I guess I could look at where the padding is?

About the padding, I do not know why it sometime is `\x00` and sometimes `\xFF`.

Regarding an issue on Gromacs, it may at least be worth asking the question on the mailing list. I wonder if they intend the TPR to be endianness specific, if they intend the TPR file to be 4 times as big as necessary, and if the content of the padding is meaningful.

Also, @ezavod, can I use your tpr files for the tests?

**Comment by jbarnoud** [comment]
>  I guess I could look at where the padding is?

Nope. Where the padding is actually depends on the XDR format that has fixed byte-order.

>  Should one raise an issue in GROMACS regarding the TPR body?

The more I look at it, the more it looks like a regression.

**Comment by ezavod** [comment]
> The more I look at it, the more it looks like a regression.

I've just asked on the GROMACS developers mailing list. I'm curious what their assessment is.
But I also think that this can't be really the aspired result.

> can I use your tpr files for the tests?

Sure, feel free to use them in any way you want. If needed you could also find your files '2lyz' and 'dummy' as big-endian [here](https://uni-muenster.sciebo.de/s/2xC9RlmhE3RxbLQ) (BTW `nsteps` is changed).

**Comment by jbarnoud** [comment]
> I've just asked on the GROMACS developers mailing list.
> Sure, feel free to use them in any way you want.

Thank you!

> As a side note: I think I've got a pretty plausible understanding on how this padding works:
At the end the whole array of chars that is buffered in gmx::InMemorySerializer is written via the FileIOXdrSerializer::doCharArray method. This writes every byte as an standard XDR int. The padding results because an int is (usually) 4 bytes and therefore 4 bytes are written instead of one.

I went that far. I was surprised that their char was using 4 bytes and not 1. The use of `doCharArray` makes me think that somebody expected `doChar` to read 1 byte at a time like some other implementations of XDR. I may be completely wrong, though.

I did not get the relation between the padding and the sign bit. That is interesting.

**Comment by acmnpv** [comment]
Hello from the GROMACS team, I just uploaded a fix for the endianess issue (https://gerrit.gromacs.org/c/gromacs/+/15059) and verified that it works for me, but please check it out as well.

Still checking the size issue

**Comment by jbarnoud** [comment]
@acmnpv Thank you for the very prompt fix! I'll implement it in here as soon as possible.

@ezavod Whenever you have a chance, could you generate a test file with this version of gromacs? Thanks.

**Comment by ezavod** [comment]
> Whenever you have a chance, could you generate a test file with this version of gromacs? 

Sure, will try this as soon as possible. But might take some time due to compile time on my old laptop.

**Comment by ezavod** [comment]
@jbarnoud I've created files that match patchset 3 (i.e. with `tpxv_AddedEndianField`).
Just in case somebody needs them they are [here](https://uni-muenster.sciebo.de/s/2xC9RlmhE3RxbLQ).
But as things are currently going, the body will be always little-endian, so this becomes an non-issue.

**Comment by richardjgowers** [comment]
I’m pretty sure people are still using gmx 5, we have to support all Tprs
not just recent ones. At least a check that detects big endianess and
aborts with a nice message.

**Comment by ezavod** [comment]
@richardjgowers The problems that I brought up here were introduced in the branch for version 2020. In version 2019 and below the TPRs were written with an XDR compliant serializer (i.e. always big-andian), so there is no difference between files from different systems (as one could easily verify with the files I initially uploaded: the headers (first few bytes) in those files are written with this XDR serializer and do not differ). I didn't explicitly looked at the source of very old versions like gmx 5, but there is nor reason to believe that they have any issue to be read (and AFAIK mda has a testsuite that verifies compatibility with everything that is newer than gmx 4.0).
As the only affected versions are beta or release candidates I would suggest to not care about files that were generated with those versions. The way its currently going this does not seem to be an issue in the 2020 release version (I'm very impressed by the fast response from the gromacs devs).

**Comment by richardjgowers** [comment]
@ezavod ah ok cool, so we've caught this in beta, that's awesome! Yeah you can close this once gmx is patched

**Comment by ezavod** [comment]
This has been fixed on the GROMACS side here: https://gerrit.gromacs.org/c/gromacs/+/15102 (see summary [here](https://mailman-1.sys.kth.se/pipermail/gromacs.org_gmx-developers/2019-December/010629.html))
TLDR: TPR body is now _always_ big-endian (as before) and the padding is gone. Those files should be easily readable with the normal XDR unpacker and no need for the fancy new implementation I guess.

**Comment by mabraham** [comment]
On behalf of the GROMACS team, thanks for the report. This made life better for everybody.

**Comment by richardjgowers** [comment]
Thanks for not changing the endianness halfway through a file, that would
have fried a few brains 😅
